Deployment of ADF from SAR fails when creating a new Deployment account

We tried to deploy ADF en let it create a new deployment account, but it failed after creating a new account. When trying to create an account the error "ConcurrentModificationException" is caught and it will retry. The second time it succeeds but it continues to run the function and the dict "create_account" is not set. This caused the deployment to fail and rolls back everything. The deployment account is being been created. 

2020-06-30T16:42:12.327+02:00` START RequestId: 81032704-4223-4fb4-abfe-a90114bf4a95 Version: $LATEST

2020-06-30T16:42:12.554+02:00 [INFO] 2020-06-30T14:42:12.553Z 81032704-4223-4fb4-abfe-a90114bf4a95 Creating account ...

2020-06-30T16:42:12.991+02:00 [INFO]	2020-06-30T14:42:12.990Z	81032704-4223-4fb4-abfe-a90114bf4a95	An error occurred (ConcurrentModificationException) when calling the CreateAccount operation: AWS Organizations can't complete your request because it conflicts with another attempt to modify the same entity. Try again later.

2020-06-30T16:42:18.025+02:00 [INFO] 2020-06-30T14:42:18.25Z 81032704-4223-4fb4-abfe-a90114bf4a95 Creating account ...

2020-06-30T16:42:18.879+02:00 [INFO] 2020-06-30T14:42:18.879Z 81032704-4223-4fb4-abfe-a90114bf4a95 Account creation requested, request ID: car-e5cafbe0badf11eaaa4712a0c7517226

2020-06-30T16:42:18.937+02:00 [INFO] 2020-06-30T14:42:18.937Z 81032704-4223-4fb4-abfe-a90114bf4a95 Account creation still in progress, waiting.. then calling again with car-e5cafbe0badf11eaaa4712a0c7517226

2020-06-30T16:42:29.006+02:00 [INFO] 2020-06-30T14:42:29.5Z 81032704-4223-4fb4-abfe-a90114bf4a95 Account created: 767015857616

2020-06-30T16:42:29.011+02:00 [ERROR] 2020-06-30T14:42:29.6Z 81032704-4223-4fb4-abfe-a90114bf4a95 Failure: Unexpected exception

2020-06-30T16:42:29.011+02:00 Traceback (most recent call last):

2020-06-30T16:42:29.011+02:00 File "/var/task/cfn_custom_resource/decorator.py", line 59, in lambda_handler

2020-06-30T16:42:29.011+02:00 phy_id, data = fn(event, context)

2020-06-30T16:42:29.011+02:00 File "/var/task/main.py", line 72, in create_

2020-06-30T16:42:29.011+02:00 existing_account_id, account_name, account_email, cross_account_access_role_name)

2020-06-30T16:42:29.011+02:00 File "/var/task/main.py", line 148, in ensure_account

2020-06-30T16:42:29.011+02:00 request_id = create_account["CreateAccountStatus"]["Id"]

2020-06-30T16:42:29.011+02:00 UnboundLocalError: local variable 'create_account' referenced before assignment

2020-06-30T16:42:29.211+02:00 END RequestId: 81032704-4223-4fb4-abfe-a90114bf4a95

2020-06-30T16:42:29.211+02:00 REPORT RequestId: 81032704-4223-4fb4-abfe-a90114bf4a95 Duration: 16883.73 ms Billed Duration: 16900 ms Memory Size: 128 MB Max Memory Used: 81 MB Init Duration: 338.50 ms

2020-06-30T16:42:46.116+02:00 START RequestId: d8c6daf5-5769-43a0-b3fe-61c37c13263b Version: $LATEST

2020-06-30T16:42:46.121+02:00 [INFO] 2020-06-30T14:42:46.121Z d8c6daf5-5769-43a0-b3fe-61c37c13263b Unrecognized physical resource: 2b371fd27d220c2a2ab1aaf352f01bf1. Assuming no delete necessary

2020-06-30T16:42:46.293+02:00 END RequestId: d8c6daf5-5769-43a0-b3fe-61c37c13263b

2020-06-30T16:42:46.293+02:00 REPORT RequestId: d8c6daf5-5769-43a0-b3fe-61c37c13263b Duration: 172.83 ms Billed Duration: 200 ms Memory Size: 128 MB Max Memory Used: 81 MB

--- a/Makefile
+++ b/Makefile
@@ -3,6 +3,7 @@ init:
 
 test:
 	# Run unit tests
+	pytest src/lambda_codebase/account -vvv -s -c src/lambda_codebase/account/pytest.ini
 	pytest src/lambda_codebase/initial_commit/bootstrap_repository -vvv -s -c src/lambda_codebase/initial_commit/bootstrap_repository/pytest.ini
 	pytest src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/lambda_codebase -vvv -s -c src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/lambda_codebase/pytest.ini
 	pytest src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python -vvv -s -c src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/pytest.ini

--- a/src/lambda_codebase/account/main.py
+++ b/src/lambda_codebase/account/main.py
@@ -29,6 +29,7 @@ ORGANIZATION_CLIENT = boto3.client("organizations")
 SSM_CLIENT = boto3.client("ssm")
 LOGGER = logging.getLogger(__name__)
 LOGGER.setLevel(logging.INFO)
+MAX_RETRIES = 120  # => 120 retries * 5 seconds = 10 minutes
 
 
 class InvalidPhysicalResourceId(Exception):
@@ -113,8 +114,8 @@ def delete_(event, _context):
 def ensure_account(existing_account_id: str,
                    account_name: str,
                    account_email: str,
-                   cross_account_access_role_name: str) -> Tuple[AccountId,
-                                                                 bool]:
+                   cross_account_access_role_name: str,
+                   no_retries: int = 0) -> Tuple[AccountId, bool]:
     # If an existing account ID was provided, use that:
     if existing_account_id:
         return existing_account_id, False
@@ -137,17 +138,22 @@ def ensure_account(existing_account_id: str,
             IamUserAccessToBilling="ALLOW",
         )
     except ORGANIZATION_CLIENT.exceptions.ConcurrentModificationException as err:
-        LOGGER.info(err)
-        time.sleep(5)
-        ensure_account(
+        return handle_concurrent_modification(
+            err,
             existing_account_id,
             account_name,
             account_email,
-            cross_account_access_role_name)
+            cross_account_access_role_name,
+            no_retries + 1,
+        )
 
     request_id = create_account["CreateAccountStatus"]["Id"]
     LOGGER.info("Account creation requested, request ID: %s", request_id)
 
+    return wait_on_account_creation(request_id)
+
+
+def wait_on_account_creation(request_id: str) -> Tuple[AccountId, bool]:
     while True:
         account_status = ORGANIZATION_CLIENT.describe_create_account_status(
             CreateAccountRequestId=request_id
@@ -157,10 +163,36 @@ def ensure_account(existing_account_id: str,
             raise Exception("Failed to create account because %s" % reason)
         if account_status["CreateAccountStatus"]["State"] == "IN_PROGRESS":
             LOGGER.info(
-                "Account creation still in progress, waiting.. then calling again with %s",
+                "Account creation still in progress, waiting.. "
+                "then calling again with %s",
                 request_id)
             time.sleep(10)
         else:
             account_id = account_status["CreateAccountStatus"]["AccountId"]
             LOGGER.info("Account created: %s", account_id)
             return account_id, True
+
+
+def handle_concurrent_modification(
+    error: Exception,
+    existing_account_id: str,
+    account_name: str,
+    account_email: str,
+    cross_account_access_role_name: str,
+    no_retries: int = 0,
+) -> Tuple[AccountId, bool]:
+    LOGGER.info("Attempt %d - hit %s", no_retries + 1, error)
+    if no_retries > MAX_RETRIES:
+        LOGGER.error(
+            "Reached maximum number of retries to create the account. "
+            "Raising error to abort the execution."
+        )
+        raise error
+    time.sleep(5)
+    return ensure_account(
+        existing_account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+        no_retries + 1,
+    )

--- a/src/lambda_codebase/account/pytest.ini
+++ b/src/lambda_codebase/account/pytest.ini
@@ -1,0 +1,2 @@
+[pytest]
+testpaths = tests

--- a/src/lambda_codebase/account/tests/__init__.py
+++ b/src/lambda_codebase/account/tests/__init__.py
@@ -1,0 +1,4 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# pylint: skip-file

--- a/src/lambda_codebase/account/tests/test_main.py
+++ b/src/lambda_codebase/account/tests/test_main.py
@@ -1,0 +1,440 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# pylint: skip-file
+
+import pytest
+from mock import patch, call
+from main import (
+    ensure_account, handle_concurrent_modification, wait_on_account_creation,
+    MAX_RETRIES,
+)
+
+
+class ConcurrentModificationException(Exception):
+    pass
+
+
+class ParameterNotFound(Exception):
+    pass
+
+
+class OtherException(Exception):
+    pass
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.SSM_CLIENT')
+@patch('main.wait_on_account_creation')
+@patch('main.handle_concurrent_modification')
+@patch('main.LOGGER')
+def test_deployment_account_given(logger, concur_mod_fn, wait_on_fn,
+                                  ssm_client, org_client):
+    account_id = "123456789012"
+    account_name = "test-deployment-account"
+    account_email = "test-deployment@fake-test.domain"
+    cross_account_access_role_name = "some-role"
+    ssm_client.exceptions.ParameterNotFound = ParameterNotFound
+    org_client.exceptions.ConcurrentModificationException = \
+        ConcurrentModificationException
+
+    returned_account_id, created = ensure_account(
+        account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+    )
+
+    assert returned_account_id == account_id
+    assert not created
+    logger.info.assert_not_called()
+    concur_mod_fn.assert_not_called()
+    wait_on_fn.assert_not_called()
+    ssm_client.get_parameter.assert_not_called()
+    org_client.create_account.assert_not_called()
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.SSM_CLIENT')
+@patch('main.wait_on_account_creation')
+@patch('main.handle_concurrent_modification')
+@patch('main.LOGGER')
+def test_deployment_account_found_with_ssm(logger, concur_mod_fn, wait_on_fn,
+                                           ssm_client, org_client):
+    account_id = "123456789012"
+    given_account_id = ""
+    account_name = "test-deployment-account"
+    account_email = "test-deployment@fake-test.domain"
+    cross_account_access_role_name = "some-role"
+    ssm_client.exceptions.ParameterNotFound = ParameterNotFound
+    org_client.exceptions.ConcurrentModificationException = \
+        ConcurrentModificationException
+
+    ssm_client.get_parameter.return_value = {
+        "Parameter": {
+            "Value": account_id,
+        }
+    }
+
+    returned_account_id, created = ensure_account(
+        given_account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+    )
+
+    assert returned_account_id != given_account_id
+    assert returned_account_id == account_id
+    assert not created
+    logger.info.assert_not_called()
+    concur_mod_fn.assert_not_called()
+    wait_on_fn.assert_not_called()
+    ssm_client.get_parameter.assert_called_once_with(
+        Name="deployment_account_id",
+    )
+    org_client.create_account.assert_not_called()
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.SSM_CLIENT')
+@patch('main.wait_on_account_creation')
+@patch('main.handle_concurrent_modification')
+@patch('main.LOGGER')
+def test_deployment_account_create_success(logger, concur_mod_fn, wait_on_fn,
+                                           ssm_client, org_client):
+    request_id = "random-request-id"
+    account_id = "123456789012"
+    given_account_id = ""
+    account_name = "test-deployment-account"
+    account_email = "test-deployment@fake-test.domain"
+    cross_account_access_role_name = "some-role"
+    ssm_client.exceptions.ParameterNotFound = ParameterNotFound
+    org_client.exceptions.ConcurrentModificationException = \
+        ConcurrentModificationException
+
+    ssm_client.get_parameter.side_effect = ParameterNotFound("Test")
+    org_client.create_account.return_value = {
+        'CreateAccountStatus': {
+            'Id': request_id,
+        }
+    }
+    wait_on_fn.return_value = (account_id, True)
+
+    returned_account_id, created = ensure_account(
+        given_account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+    )
+
+    assert returned_account_id != given_account_id
+    assert returned_account_id == account_id
+    assert created
+    logger.info.assert_has_calls([
+        call("Creating account ..."),
+        call("Account creation requested, request ID: %s", request_id),
+    ])
+    concur_mod_fn.assert_not_called()
+    wait_on_fn.assert_called_once_with(request_id)
+    ssm_client.get_parameter.assert_called_once_with(
+        Name="deployment_account_id",
+    )
+    org_client.create_account.assert_called_once_with(
+        Email=account_email,
+        AccountName=account_name,
+        RoleName=cross_account_access_role_name,
+        IamUserAccessToBilling="ALLOW",
+    )
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.SSM_CLIENT')
+@patch('main.wait_on_account_creation')
+@patch('main.handle_concurrent_modification')
+@patch('main.LOGGER')
+def test_deployment_account_create_failed_concur(logger, concur_mod_fn,
+                                                 wait_on_fn, ssm_client,
+                                                 org_client):
+    request_id = "random-request-id"
+    account_id = "123456789012"
+    given_account_id = ""
+    account_name = "test-deployment-account"
+    account_email = "test-deployment@fake-test.domain"
+    cross_account_access_role_name = "some-role"
+    ssm_client.exceptions.ParameterNotFound = ParameterNotFound
+    org_client.exceptions.ConcurrentModificationException = \
+        ConcurrentModificationException
+
+    ssm_client.get_parameter.side_effect = ParameterNotFound("Test")
+    org_client.create_account.side_effect = ConcurrentModificationException(
+        "Test",
+    )
+    concur_mod_fn.return_value = (account_id, True)
+
+    returned_account_id, created = ensure_account(
+        given_account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+    )
+
+    assert returned_account_id != given_account_id
+    assert returned_account_id == account_id
+    assert created
+    logger.info.assert_has_calls([
+        call("Creating account ..."),
+    ])
+    concur_mod_fn.assert_called_once_with(
+        org_client.create_account.side_effect,
+        given_account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+        1,
+    )
+    wait_on_fn.assert_not_called()
+    ssm_client.get_parameter.assert_called_once_with(
+        Name="deployment_account_id",
+    )
+    org_client.create_account.assert_called_once_with(
+        Email=account_email,
+        AccountName=account_name,
+        RoleName=cross_account_access_role_name,
+        IamUserAccessToBilling="ALLOW",
+    )
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.SSM_CLIENT')
+@patch('main.wait_on_account_creation')
+@patch('main.handle_concurrent_modification')
+@patch('main.LOGGER')
+def test_deployment_account_create_failed_other(logger, concur_mod_fn,
+                                                wait_on_fn, ssm_client,
+                                                org_client):
+    request_id = "random-request-id"
+    account_id = "123456789012"
+    given_account_id = ""
+    account_name = "test-deployment-account"
+    account_email = "test-deployment@fake-test.domain"
+    cross_account_access_role_name = "some-role"
+    correct_error_message = "Some other exception"
+    ssm_client.exceptions.ParameterNotFound = ParameterNotFound
+    org_client.exceptions.ConcurrentModificationException = \
+        ConcurrentModificationException
+
+    ssm_client.get_parameter.side_effect = ParameterNotFound("Test")
+    org_client.create_account.side_effect = \
+        OtherException(correct_error_message)
+
+    with pytest.raises(OtherException) as excinfo:
+        ensure_account(
+            given_account_id,
+            account_name,
+            account_email,
+            cross_account_access_role_name,
+        )
+
+    error_message = str(excinfo.value)
+    assert error_message.find(correct_error_message) >= 0
+
+    logger.info.assert_has_calls([
+        call("Creating account ..."),
+    ])
+    concur_mod_fn.assert_not_called()
+    wait_on_fn.assert_not_called()
+    ssm_client.get_parameter.assert_called_once_with(
+        Name="deployment_account_id",
+    )
+    org_client.create_account.assert_called_once_with(
+        Email=account_email,
+        AccountName=account_name,
+        RoleName=cross_account_access_role_name,
+        IamUserAccessToBilling="ALLOW",
+    )
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.time')
+@patch('main.LOGGER')
+def test_deployment_account_wait_exception(logger, time_mock, org_client):
+    request_id = "random-request-id"
+    correct_error_message = "Failed test"
+    org_client.describe_create_account_status.side_effect = \
+        OtherException(correct_error_message)
+
+    with pytest.raises(OtherException) as excinfo:
+        wait_on_account_creation(
+            request_id,
+        )
+
+    error_message = str(excinfo.value)
+    assert error_message.find(correct_error_message) >= 0
+
+    logger.info.assert_not_called()
+    time_mock.sleep.assert_not_called()
+    org_client.describe_create_account_status.assert_called_once_with(
+        CreateAccountRequestId=request_id,
+    )
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.time')
+@patch('main.LOGGER')
+def test_deployment_account_wait_on_failed(logger, time_mock, org_client):
+    request_id = "random-request-id"
+    failure_reason = "The failure on create status request"
+    org_client.describe_create_account_status.return_value = {
+        "CreateAccountStatus": {
+            "State": "FAILED",
+            "FailureReason": failure_reason,
+        }
+    }
+    correct_error_message = \
+        "Failed to create account because %s" % failure_reason
+
+    with pytest.raises(Exception) as excinfo:
+        wait_on_account_creation(
+            request_id,
+        )
+
+    error_message = str(excinfo.value)
+    assert error_message.find(correct_error_message) >= 0
+
+    logger.info.assert_not_called()
+    time_mock.sleep.assert_not_called()
+    org_client.describe_create_account_status.assert_called_once_with(
+        CreateAccountRequestId=request_id,
+    )
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.time')
+@patch('main.LOGGER')
+def test_deployment_account_wait_in_progress_success(logger, time_mock, org_client):
+    request_id = "random-request-id"
+    account_id = "123456789012"
+    org_client.describe_create_account_status.side_effect = [
+        {
+            "CreateAccountStatus": {
+                "State": "IN_PROGRESS",
+            }
+        },
+        {
+            "CreateAccountStatus": {
+                "State": "SUCCEEDED",
+                "AccountId": account_id,
+            }
+        },
+    ]
+
+    returned_account_id, created = wait_on_account_creation(
+        request_id,
+    )
+
+    assert returned_account_id == account_id
+    assert created
+
+    logger.info.assert_has_calls([
+        call(
+            "Account creation still in progress, waiting.. "
+            "then calling again with %s",
+            request_id,
+        ),
+        call("Account created: %s", account_id)
+    ])
+    time_mock.sleep.assert_called_once_with(10)
+    org_client.describe_create_account_status.assert_has_calls([
+        call(CreateAccountRequestId=request_id),
+        call(CreateAccountRequestId=request_id),
+    ])
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.ensure_account')
+@patch('main.time')
+@patch('main.LOGGER')
+def test_deployment_account_handle_concurrent_last_try(logger, time_mock,
+                                                    ensure_fn, org_client):
+    account_id = "123456789012"
+    given_account_id = ""
+    account_name = "test-deployment-account"
+    account_email = "test-deployment@fake-test.domain"
+    cross_account_access_role_name = "some-role"
+    no_retries = MAX_RETRIES
+    org_client.exceptions.ConcurrentModificationException = \
+        ConcurrentModificationException
+    correct_error_message = "SomeError"
+    err = ConcurrentModificationException(correct_error_message)
+
+    ensure_fn.return_value = (account_id, True)
+
+    returned_account_id, created = handle_concurrent_modification(
+        err,
+        given_account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+        no_retries,
+    )
+
+    assert returned_account_id == account_id
+    assert created
+
+    logger.info.assert_called_once_with(
+        "Attempt %d - hit %s", no_retries + 1, err,
+    )
+    logger.error.assert_not_called()
+    time_mock.sleep.assert_called_once_with(5)
+    ensure_fn.assert_called_once_with(
+        given_account_id,
+        account_name,
+        account_email,
+        cross_account_access_role_name,
+        no_retries + 1,
+    )
+
+
+@patch('main.ORGANIZATION_CLIENT')
+@patch('main.ensure_account')
+@patch('main.time')
+@patch('main.LOGGER')
+def test_deployment_account_handle_concurrent_too_many_tries(logger,
+                                                             time_mock,
+                                                             ensure_fn,
+                                                             org_client):
+    account_id = "123456789012"
+    given_account_id = ""
+    account_name = "test-deployment-account"
+    account_email = "test-deployment@fake-test.domain"
+    cross_account_access_role_name = "some-role"
+    no_retries = MAX_RETRIES + 1
+    org_client.exceptions.ConcurrentModificationException = \
+        ConcurrentModificationException
+    correct_error_message = "SomeError"
+    err = ConcurrentModificationException(correct_error_message)
+
+    ensure_fn.return_value = (account_id, True)
+
+    with pytest.raises(ConcurrentModificationException) as excinfo:
+        handle_concurrent_modification(
+            err,
+            given_account_id,
+            account_name,
+            account_email,
+            cross_account_access_role_name,
+            no_retries,
+        )
+
+    error_message = str(excinfo.value)
+    assert error_message.find(correct_error_message) >= 0
+
+    logger.info.assert_called_once_with(
+        "Attempt %d - hit %s", no_retries + 1, err,
+    )
+    logger.error.assert_called_once_with(
+        "Reached maximum number of retries to create the account. "
+        "Raising error to abort the execution."
+    )
+    time_mock.sleep.assert_not_called()
+    ensure_fn.assert_not_called()

--- a/tox.ini
+++ b/tox.ini
@@ -29,6 +29,7 @@ deps =
     -r{toxinidir}/requirements.txt
     -r{toxinidir}/src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/requirements.txt
     -r{toxinidir}/src/lambda_codebase/initial_commit/requirements.txt
+    -r{toxinidir}/src/lambda_codebase/account/requirements.txt
 commands =
     make test
     make lint

